### PR TITLE
feat: add flag to view process tree

### PR DESCRIPTION
**Pull request by snprajwal**
CRIT already allows us to explore the process tree, file descriptors, memory map, and RSS map for a checkpoint. These features can be integrated with checkpointctl to allow more info to be displayed with `checkpointctl show`. This PR introduces `checkpointctl show --ps-tree` for viewing the process tree.

Here's a sample output of the tree displayed for a single process:

```
Process tree

Container
└── [1]  piggie
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -13,16 +13,17 @@ require (
 require (
 	github.com/docker/go-units v0.5.0 // indirect
 	github.com/inconshreveable/mousetrap v1.1.0 // indirect
-	github.com/klauspost/compress v1.16.4 // indirect
-	github.com/klauspost/pgzip v1.2.5 // indirect
-	github.com/mattn/go-runewidth v0.0.9 // indirect
+	github.com/klauspost/compress v1.16.5 // indirect
```
Comment (rst0git) [inline go.mod:16]: @snprajwal Could you move this patch in a separate pull request?

**Comment by snprajwal** [comment]
CRIU dump and restore are failing in the CI, I'm not sure why. Is it because of insufficient permissions or something? I based the workflow directly off of the one we have in go-criu.

**Comment by adrianreber** [comment]
> CRIU dump and restore are failing in the CI, I'm not sure why. Is it because of insufficient permissions or something? I based the workflow directly off of the one we have in go-criu.

We run in a unprivileged container. GitHub actions supports privileged containers. Please try that.

**Comment by snprajwal** [comment]
> We run in a unprivileged container. GitHub actions supports privileged containers. Please try that.

Looks like that helped - now both workflows fail on restore :laughing:

**Comment by adrianreber** [comment]
> > We run in a unprivileged container. GitHub actions supports privileged containers. Please try that.
> 
> Looks like that helped - now both workflows fail on restore :laughing: 

You either need to print the logs in an error case or log to stdout.

It could be necessary to mount /lib/modules from the host into the container so that necessary module can be loaded.

But without logs it is difficult to tell.

**Comment by snprajwal** [comment]
I don't want to create too much noise on this PR. Let me try it out on a separate branch, and push the changes here once I get it working

**Comment by adrianreber** [comment]
I think we might need to wait a bit with this PR. My favourite approach right now would be to make everything PID tree based. Maybe the table view is only useful for the quick overview. I am curious in how far the tree based approach would work for `checkpointctl` overall.

**Comment by adrianreber** [comment]
We talked about binary sizes before. I think we should include a CI check about the binary size like Podman has. Build each PR without the PR and record the binary size, build the PR with the changes and record the binary size. Compare both sizes and fail if +10%.

**Comment by snprajwal** [comment]
I've added the process tree logic here. I didn't want to create a separate branch and PR for it as of now since I had important changes in the first two commits, and did not want to duplicate them. If this looks good, we can discuss how file descriptors, memory maps, etc. integrate into this and add them accordingly. I still have the commit and code for the file descriptors feature saved locally.

**Comment by adrianreber** [comment]
There are a couple of things that should be done differently. This PR mixes too many things. There is some renaming going on which should be a separate commit. The changes to the testing and the new features should be different PRs.

For the new feature I would like to see some example output in the PR description.

Instead of the loop.c example, let's try to use `piggie.c` from go-criu, which also creates a couple of namespaces as this will be helpful at some point during testing.

Also the binary size comparison PR should be done before any more includes from go-criu.

**Comment by snprajwal** [comment]
Got it, I'll split it up and open separate PRs.

File: checkpointctl.go [hunk 2]
```diff
@@ -47,22 +48,28 @@ func setupShow() *cobra.Command {
 	}
 	flags := cmd.Flags()
 	flags.BoolVar(
-		&printStats,
+		&stats,
 		"print-stats",
 		false,
-		"Print checkpointing statistics if available",
+		"Print checkpointstatistics if available",
```
Comment (rst0git) [inline checkpointctl.go:54]: "checkpoint statistics" are two words

Comment (snprajwal) [reply checkpointctl.go:54]: Ack I think I messed up while rebasing. I'll fix it when I rebase after #61 is merged.

**Comment by snprajwal** [comment]
@adrianreber I've updated the PR description with the tree we currently display during tests (with piggie).

**Comment by adrianreber** [comment]
Looks good for a first step. I like it.

File: container.go [hunk 3]
```diff
@@ -117,12 +118,23 @@ func showContainerCheckpoint(checkpointDirectory, input string) error {
...
 		if err != nil {
-			return fmt.Errorf("unable to display checkpointing statistics: %w", err)
+			return fmt.Errorf("failed to get dump statistics: %w", err)
 		}
 
 		renderDumpStats(dumpStats)
 	}
 
+	if psTree {
+		c := crit.New("", "", checkpointDirectory, false, false)
```
Comment (rst0git) [inline container.go:128]: > Error: failed to get process tree: Error opening binary file: open /tmp/checkpointctl1506966507/pstree.img: no such file or directory

Comment (snprajwal) [reply container.go:128]: Ack, figured this out. The temporary directory being created is one level above the checkpoint directory. I need to append that path here.

Comment (snprajwal) [reply container.go:128]: I tried listing the files present in the temporary directory, and got this:
```
/tmp/checkpointctl65800022/config.dump
/tmp/checkpointctl65800022/spec.dump
/tmp/checkpointctl65800022/stats-dump
```

File: checkpointctl.go [hunk 4]
```diff
@@ -122,7 +130,9 @@ func show(cmd *cobra.Command, args []string) error {
 		}
 	}()
 
-	if err := untarFiles(input, dir, metadata.SpecDumpFile, metadata.ConfigDumpFile); err != nil {
+	// List of files required from the checkpoint archive
+	files := []string{metadata.SpecDumpFile, metadata.ConfigDumpFile, metadata.CheckpointDirectory}
```
Comment (adrianreber) [inline checkpointctl.go:134]: This works, but this is against the goal I had in mind with the change to only unpack the needed files.

If the user requests the process tree, only then should the corresponding files be unpacked. According to the test case below only `pstree.img` and `core-*.img` files are needed.

File: container.go [hunk 5]
```diff
@@ -257,38 +295,15 @@ func shortenPath(path string) string {
...
-
-	// Remove files from ExcludePatterns if they are present
-	for _, file := range files {
-		for i, pattern := range options.ExcludePatterns {
-			if pattern == file {
-				options.ExcludePatterns = append(options.ExcludePatterns[:i], options.ExcludePatterns[i+1:]...)
-				break
-			}
-		}
+		IncludeFiles: files,
```
Comment (adrianreber) [inline container.go:306]: I understand what you are trying to do, but I am not sure it actually works as expected. Did you try it out?

When I last used this library, `IncludeFiles` did not work as I expected it during unpacking.

Comment (snprajwal) [reply container.go:306]: Last I checked, it worked as expected. Let me double check once more.

File: checkpointctl.go (hunk 4 above)
Comment (snprajwal) [reply checkpointctl.go:134]: Got it, I'll change this. But do note that we will have to add more files as we go along, for the other explore types.

Comment (snprajwal) [reply checkpointctl.go]: I have changed it to include the individual files that each feature requires. For now, we are only unarchiving `pstree.img` and `core-*.img` for the process tree. A nifty feature of `IncludeFiles` with `tar` is that it allows us to specify wildcards!

File: test/checkpointctl.bats [hunk 6]
```diff
@@ -113,7 +113,7 @@ function teardown() {
 	( cd "$TEST_TMP_DIR1" && tar cf "$TEST_TMP_DIR2"/test.tar . )
 	checkpointctl show "$TEST_TMP_DIR2"/test.tar --stats
 	[ "$status" -eq 1 ]
-	[[ ${lines[6]} == *"unable to display checkpointing statistics"* ]]
+	[[ ${lines[6]} == *"failed to get dump statistics"* ]]
```
Comment (rst0git) [inline test/checkpointctl.bats:116]: Is this change intentional? (It looks unrelated)

File: container.go [hunk 7]
```diff
@@ -221,6 +228,31 @@ func renderDumpStats(dumpStats *images.DumpStatsEntry) {
 	table.Render()
 }
 
+func renderPsTree(psTree *crit.PsTree, containerName string) {
+	var tree treeprint.Tree
+	if containerName != "" {
+		tree = treeprint.NewWithRoot(containerName)
+	} else {
+		tree = treeprint.NewWithRoot("Container")
+	}
```
Comment (rst0git) [inline container.go:237]: Could we simplify this condition to something like the following?
```Go
if containerName == "" {
	containerName = "Container"
}
tree = treeprint.NewWithRoot(containerName)
```

**Review by rst0git — APPROVED** [review]

File: container.go (hunk 5 above)
Comment (snprajwal) [reply container.go:299]: Thanks for pointing this out. I realised now that during tests, we move only the required files into the checkpoint folder before creating the archive. Since I used that to test this functionality, it appeared as though only the necessary files are being unpacked. I'll fix this.

**Review by adrianreber — APPROVED** [review]
